hydrate/match.py

```diff
@@ -0,0 +1,193 @@
+
+            if PartialMatchResults.partial_matches:
+                self.match_categories.append(
+                    MatchCategory(top_comment=PARTIAL_MATCH_COMMENT,
+                                  start_index=self.get_start_index(),
+                                  matches=PartialMatchResults.partial_matches))
+
+            if PartialMatchResults.leftovers:
+                NoMatchResults = self.get_no_matches(PartialMatchResults.leftovers)
+
+                self.match_categories.append(
+                    MatchCategory(top_comment=NO_MATCH_COMMENT,
+                                  start_index=self.get_start_index(),
+                                  matches=NoMatchResults))
+
+        return self.match_categories
+
+    def get_start_index(self):
+        """Calculate start_index based on the last MatchCategory in the list."""
+        start_idx = 0
+        if self.match_categories:
+            start_idx = len(self.match_categories[-1].matches) \
+                        + self.match_categories[-1].start_index
+        return start_idx
+
+    def get_full_matches(self, cluster_components):
+        """Determine which components fully match the cluster.
+
+        Returns:
+            FullMatchResults: namedtuple
+                full_matches: list of components or []
+                leftovers: list of components or None
+
+        """
+        FullMatchResults = namedtuple('FullMatchResults', ['full_matches', 'leftovers'])
+        full_matches = []
+        leftovers = None
+        cluster_map = {cc.name: cc for cc in cluster_components}
+        cluster_set = set(cluster_map.keys())
+
+        for rc in self.repo_components:
+            repo_set = set(rc.name.split('-'))
+            if repo_set.issubset(cluster_set):
+                cc_names = cluster_set.intersection(repo_set)
+                cluster_matches = [cluster_map[component_name] for component_name in cc_names]
+                cluster_set.difference_update(repo_set)
+
+                full_matches.append(
+                    Match(category=MatchCategory.FULL_MATCH,
+                          cluster_matches=cluster_matches,
+                          repo_match=rc))
+
+        if cluster_set:
+            leftovers = [cc for cc in cluster_components if cc.name in cluster_set]
+
+        return FullMatchResults(full_matches, leftovers)
+
+    def get_partial_matches(self, full_match_leftovers):
+        """Determine which components partially match the cluster.
+
+        Args:
+            repo_components: list of Components from the definitions repository
+            full_match_leftovers: list of leftover deployments after get_full_matches
+
+        Returns:
+            PartialMatchResults: namedtuple
+                partial_matches: dict(key: component.name, value: list of components)
+                leftovers: list of components
+
+        """
+        PartialMatchResults = namedtuple('PartialMatchResults', ['partial_matches', 'leftovers'])
```

Comment:
leftovers -> no_matches?

Reply:
Yes, the leftovers get passed into the get_no_matches method. I suppose I could rename the leftovers to no_matches.

Reply:
Resolved in commit https://github.com/microsoft/hydrate/pull/26/commits/02df3ef98280187f991c39280c01e81c2ee9cc2a